src/Sylius/Behat/Context/Api/Admin/ManagingCountriesContext.php

```diff
@@ -177,9 +227,26 @@ public function iDoNotSpecifyTheProvince(): void
      */
     public function iRemoveProvinceName(ProvinceInterface $province): void
     {
-        $this->client->buildUpdateRequest(Resources::PROVINCES, $province->getCode());
-        $this->client->addRequestData('name', '');
-        $this->client->update();
+        $showRequest = $this->requestFactory->show(
+            PathPrefixes::ADMIN_PREFIX,
+            Resources::PROVINCES,
+            $province->getCode(),
+            'Authorization',
+            $this->sharedStorage->get('token')
+        );
+        $showResponse = $this->client->executeCustomRequest($showRequest);
+
+        $updateRequestBuilder = $this->requestBuilderFactory->update(
+            PathPrefixes::ADMIN_PREFIX,
+            Resources::PROVINCES,
+            $province->getCode(),
+            'Authorization',
+            $this->sharedStorage->get('token')
+        );
+        $updateRequestBuilder->withContent(json_decode($showResponse->getContent(), true));
+        $updateRequestBuilder->withPartialContent('name', '');
+
+        $this->client->executeCustomRequest($updateRequestBuilder->build());
```

Comment:
I'm convinced that the stuff you are proposing is a nice improvement and a good direction for it, however, I'm concerned about the chosen path. 

 This is quite a large refactor at once if we want to switch all our implementations to the new structure. What would you say, about keeping quite a similar API for that we had on client, but moving it to RequestFactory. E.g.
```php
        $this->requestFactory->update(Resources::PROVINCES, $province->getCode()); // Would request for show of give province and initiate content for update with it, then will store it in shared storage or builder is returned
        $this->requestFactory->addRequestData('name', '');  // mutates data hold in shared storage
        $request = $this->requestFactory->create(); // builds requests and returns it back

        $this->client->executeCustomRequest($request);
```

The advantage of it would be, that it simplifies the migration process, as we would already have client and request creation decoupled and both services would be stateless. The state of request will be held and maintained in shared storage. In addition if one would like to change something with builder, it would be available from shared storage

Reply:
I agree that it is quite a bit of refactor, and we probably should look for a little more approachable solution.

Yet, I can't agree with the move of `Shared Storage` into the `Request Factory`, as it will just move the pain/problem we are trying to heal. It is true that we need some sort of helper class (or are we?), but I would stick to the single responsibility of the services/classes. Isn't `addRequestData` gonna create state within `Request Factory`?

Reply:
I can agree, that we are only moving the problem away. But it is still beneficial in my opinion - we would have the client decoupled already. So instead of a stateful object with broken boundaries, we would have a stateful object and an object with proper boundaries. The next step of refactoring would be to place the state in a separate service(`SharedStorage`) to push it to a side of our processing logic. Later, the question remains, how would we provide an already created request object. Are we leaving `SharedStorage` in place or not? We can always provide it as an argument and keep the reference to it in context (a little bit similar to what SymfonyForm type is doing)